src/output_mqtt.c

```diff
@@ -0,0 +1,434 @@
+    char address[253 + 6 + 1]; // dns max + port
+    char client_id[256];
+    uint16_t message_id;
+} mqtt_client_t;
+
+static void mqtt_client_event(struct mg_connection *nc, int ev, void *ev_data)
+{
+    // note that while shutting down the ctx is NULL
+    mqtt_client_t *ctx = (mqtt_client_t *)nc->mgr->user_data;
+    // only valid in MG_EV_MQTT_ events
+    struct mg_mqtt_message *msg = (struct mg_mqtt_message *)ev_data;
+
+    //if (ev != MG_EV_POLL)
+    //    fprintf(stderr, "MQTT user handler got event %d\n", ev);
+
+    switch (ev) {
+    case MG_EV_CONNECT: {
+        int connect_status = *(int *)ev_data;
+        if (connect_status == 0) {
+            // Success
+            fprintf(stderr, "MQTT Connected...\n");
+            mg_set_protocol_mqtt(nc);
+            mg_send_mqtt_handshake_opt(nc, ctx->client_id, ctx->opts);
+        }
+        else {
+            // Error, print only once
+            if (ctx && ctx->prev_status != connect_status)
+                fprintf(stderr, "MQTT connect error: %s\n", strerror(connect_status));
+        }
+        if (ctx)
+            ctx->prev_status = connect_status;
+        break;
+    }
+    case MG_EV_MQTT_CONNACK:
+        if (msg->connack_ret_code != MG_EV_MQTT_CONNACK_ACCEPTED) {
+            fprintf(stderr, "MQTT Connection error: %d\n", msg->connack_ret_code);
+        }
+        else {
+            fprintf(stderr, "MQTT Connection established.\n");
+        }
+        break;
+    case MG_EV_MQTT_PUBACK:
+        fprintf(stderr, "MQTT Message publishing acknowledged (msg_id: %d)\n", msg->message_id);
+        break;
+    case MG_EV_MQTT_SUBACK:
+        fprintf(stderr, "MQTT Subscription acknowledged.\n");
+        break;
+    case MG_EV_MQTT_PUBLISH: {
+        fprintf(stderr, "MQTT Incoming message %.*s: %.*s\n", (int)msg->topic.len,
+                msg->topic.p, (int)msg->payload.len, msg->payload.p);
+        break;
+    }
+    case MG_EV_CLOSE:
+        if (!ctx)
+            break; // shuttig down
+        if (ctx->prev_status == 0)
+            fprintf(stderr, "MQTT Connection failed...\n");
+        // reconnect
+        if (mg_connect(nc->mgr, ctx->address, mqtt_client_event) == NULL) {
+            fprintf(stderr, "MQTT connect(%s) failed\n", ctx->address);
+        }
+        break;
+    }
+}
+
+static struct mg_mgr *mqtt_client_init(char const *host, char const *port, char const *user, char const *pass, char const *client_id)
+{
+    struct mg_mgr *mgr = calloc(1, sizeof(*mgr));
+    if (!mgr) {
+        fprintf(stderr, "calloc() failed in %s() %s:%d\n", __func__, __FILE__, __LINE__);
+        exit(1);
+    }
+
+    mqtt_client_t *ctx = calloc(1, sizeof(*ctx));
+    if (!ctx) {
+        fprintf(stderr, "calloc() failed in %s() %s:%d\n", __func__, __FILE__, __LINE__);
+        exit(1);
+    }
+    ctx->opts.user_name = user;
+    ctx->opts.password  = pass;
+    //ctx->opts.keepalive = 20;
+    //ctx->qos = 1;
+    //ctx->retain = 0; // maybe 1 after testing
```

Comment:
I guess you probably want to have both qos and retain user-configurable

Reply:
Does not need to be in this pull request though

Reply:
Good suggestion. But advanced MQTT options will have to wait a bit tough, I'll make a note.

Reply:
I too would like to see 'retain' be user configurable, with home assistant its nice to have binary sensors have a state when they start up.  

Additionally, this may be a bit of a big ask, but I would love to see a device filter mechanism of some sort, for example, say you are monitoring Interlogix security sensors, it would be nice to only events sent from a whitelist of devices, ignoring your neighbors sensors. 

Thanks again!

Reply:
I'll look into `retain` options.
Filtering may be an add-on some day. For now you could use a Python script to filter things and reinsert into MQTT.